PyAlbany/python/RandomizedCompression.py

```diff
@@ -0,0 +1,337 @@
+# @HEADER
+
+# This file contains routines for randomized truncated singular value decompositions
+# of a matrix free operator (Op), by singlePass and doublePass algorithms.
+# A routine is available for the hierarchical off-diagonal low-rank (HODLR) decomposition
+# of a matrix free operator (Op). 
+#
+# See "Finding structure with randomness, probabilistic algorithms for constructing approximate
+# matrix decompositions", N. Halko, P.G. Martinsson, J.A. Tropp SIAM Review (2011),
+# "Compressing rank-structured matrices via randomized sampling", P.G. Martinsoon, SIAM Journal
+# on Scientific Computing (2016) and "Hierarchical off-diagonal low-rank approximation of
+# Hessians in inverse problems", T. Hartland, G. Stadler, M. Perego, K. Liegeois, N. Petra (to be submitted).  
+#
+# Various helper functions are utilized to allow for a parallel implementation of the HODLR compression
+# algorithm, such as Hpartition, MPIpartition and HidxMPIproc
+#
+#
+#
+# Last updated on April 21st, 2022
+# Author Tucker Hartland, University of California, Merced
+
+
+
+from PyTrilinos import Tpetra
+from PyTrilinos import Teuchos
+try:
+    from PyAlbany import wpyalbany as wpa
+except:
+    import wpyalbany as wpa
+
+try:
+    from PyAlbany import Utils as utils
+except:
+    import Utils as utils
+
+import numpy as np
+import scipy.linalg as spla
+
+
+"""
+  singlePass: randomized singular value decomposition of a symmetric operator
+  input: Op, operator which contains a Map member data and a 
+                  dot method which takes a Tpetra MultiVector
+                             and returns a Tpetra MultiVector 
+          k, the rank of the output operator U \Lambda U^T
+  output: \Lambda, U
+          \Lambda, eigenvalues,  a nondistributed numpy array with k entries
+          U,       eigenvectors, a distributed Tpetra MultiVector with k columns
+"""
+def singlePass(Op, k, comm=Teuchos.DefaultComm.getComm()):
+    rank       = comm.getRank()
+    nprocs     = comm.getSize()
+    nElems     = Op.Map.getLocalNumElements()
+    N          = Op.Map.getGlobalNumElements()
+    # orthogonalize no more than 50 vectors at a time
+    nMaxOrthog = min(50, k)
+
+    omega = Tpetra.MultiVector(Op.Map, k, dtype="d")
+    q     = Tpetra.MultiVector(Op.Map, k, dtype="d")
+    
+    for i in range(k):
+        omega[i, :] = np.random.randn(nElems)
+    y = Op.dot(omega)
+    
+    q[:, :] = y[:, :]
+    wpa.orthogTpMVecs(q, nMaxOrthog)
+
+    C = utils.innerMVector(q, omega, comm)
+    D = utils.innerMVector(q, y, comm)
+
+    A = C.dot(C.T)
+    F = D.dot(C.T) + C.dot(D.T)
+
+    B = spla.solve_sylvester(A, A, F)
+
+    B = 0.5 * (B + B.T)
+
+    lam, utilde = np.linalg.eigh(B)
+   
+    args = np.argsort(lam)
+
+    lam    = lam[args[::-1]]
+    utilde = utilde[:, args[::-1]]
+
+    u      = utils.innerMVectorMat(q, utilde)
+    return lam, u
+
+def doublePass(Op, k, comm = Teuchos.DefaultComm.getComm(), symmetric=True):
```

Comment:
Sorry, I just noticed this. I think it's better not to provide a default value for "symmetric" (which might require changing the order of the parameters). However, if you want to give a default value, it should be `false`, so that it will work for all matrices even if not optimally.

Reply:
This default has been changed.